ts/boot/migrateToPagination.ts

```diff
@@ -0,0 +1,66 @@
+import * as Either from "fp-ts/lib/Either";
+
+import {
+  messagesStatusSelector,
+  MessageStatus
+} from "../store/reducers/entities/messages/messagesStatus";
+import { Store } from "../store/actions/types";
+import { removeMessages } from "../store/actions/messages";
+
+type Failure = {
+  error: unknown;
+  messageId: string;
+};
+
+type MigrationResult = {
+  failed: Array<Failure>;
+  succeeded: Array<string>;
+};
+
+export default async function init(
+  store: Store,
+  upsert: (id: string, messageStatus: MessageStatus) => Promise<void>
+): Promise<MigrationResult> {
+  const data = messagesStatusSelector(store.getState());
+  const allIds = Object.keys(data);
+  if (allIds.length < 1) {
+    return { failed: [], succeeded: [] };
+  }
+  // TODO: dispatch to show the loader
+
+  const requests: Array<Promise<Either.Either<Failure, string>>> = allIds.map(
+    async id => {
+      const messageStatus = data[id];
+      if (messageStatus) {
+        try {
```

Comment:
It's ⚠️ **opinionated** ⚠️, but here you could avoid the `try...catch` using the `TaskEither.tryCatch` function from `fp-ts` in this way:

```ts
return TE.tryCatch(
  () => upsert(id, messageStatus),
  error => ({
    error,
    messageId: id
  })
)
  .map(() => id)
  .run();
```

Where `TE` is `import * as TE from "fp-ts/lib/TaskEither";`.

This will force you to create a mock which is actually a `Promise` (thus reflecting the types in the code) so you'd have to change the `beforeEach` in the test to:

```ts
afterEach(() => {
  mockUpsert.mockReset();
  mockUpsert.mockReturnValue(Promise.resolve());
});
```

and the override on line `75` to:

```ts
mockUpsert.mockImplementation(async id => {
  if (id === "B") {
    throw new Error("Network error");
  }
})
```

What do you think? 🙂

Reply:
That was my first implementation indeed 😄 but I don't see any added value besides the increased complexity.

Reply:
Gotcha, I can see your point 🙂